src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala

```diff
@@ -1,41 +1,95 @@
 package edg_ide.runner
 
-import com.lowagie.text.{Document, Element, HeaderFooter, Rectangle}
-import com.lowagie.text.pdf.PdfWriter
+import com.lowagie.text.{Anchor, Document, Element, Font, FontFactory, HeaderFooter, Paragraph, Phrase, Rectangle}
+import com.lowagie.text.pdf.{PdfPCell, PdfPTable, PdfWriter}
+import edg.EdgirUtils.SimpleLibraryPath
 import edg.wir.ProtoUtil.BlockProtoToSeqMap
 import edgir.elem.elem.{BlockLike, HierarchyBlock}
 import org.eclipse.elk.graph.ElkNode
 import edg.wir.DesignPath
 import edg_ide.edgir_graph.HierarchyGraphElk.PropertyMapper
 import edg_ide.edgir_graph.{EdgeWrapper, HierarchyGraphElk, NodeDataWrapper, PortWrapper}
 import edg_ide.swing.blocks.ElkNodePainter
+import edgir.ref.ref.LibraryPath
 
 import java.awt.Color
 import java.io.{FileNotFoundException, FileOutputStream, IOException}
-
+import scala.collection.mutable
 
 object PDFGeneratorUtil{
 
+  /*
+    Note for future development:
+      If infinite loop (and eventually StackOverflow) occurs when adding header/footer, check that the
+      page size is large enough/page margin is small enough so that header/footer content do not exceed
+      the available space in the document.
+   */
+  final val PAGE_MARGIN = 15
+  final val font = new Font(Font.HELVETICA, 12f, Font.UNDERLINE, Color.BLUE)
+  final val TABLE_ROW_HEIGHT = 23f
+
+
   private def generatePageSize(node: ElkNode): (Float, Float) = {
-    val width = node.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
-    val height = node.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
+    val width = node.getWidth.toFloat + 2.5f * ElkNodePainter.margin.toFloat
+    val height = node.getHeight.toFloat + 2.5f * ElkNodePainter.margin.toFloat
     (width, height)
   }
 
+
+  def getDuplicationList(block: HierarchyBlock, path: DesignPath = DesignPath(),
+                         dupList: mutable.Map[LibraryPath, Set[DesignPath]]): Unit ={
+
+    val className = block.getSelfClass
+    dupList.getOrElseUpdate(className, Set(path)) match {
+      case existingSet: Set[DesignPath] => dupList(className) = existingSet + path
+    }
+
+    block.blocks.asPairs.map {
+      case (name, subblock) => (name, subblock.`type`)
+    }.collect {
+      case (name, BlockLike.Type.Hierarchy(subblock)) =>  (path + name, subblock)
+    }.foreach {
+      case (path, subblock) => getDuplicationList(subblock, path, dupList)
+    }
+  }
+
+
   def generate(content: HierarchyBlock,
                mappers: Seq[PropertyMapper[NodeDataWrapper, PortWrapper, EdgeWrapper]] = Seq(),
                fileName: String): Unit = {
+
+    val dupList = mutable.Map[LibraryPath, Set[DesignPath]]()
+    getDuplicationList(content, dupList = dupList)
+
     try {
```

Comment:
Not strictly related to this PR: I wouldn't wrap this in a try, instead I'd just have those errors fail noisily, and if the upper layer needs graceful error handling it should be the responsibility of that layer.

Reply:
Hmmm correct me if I'm wrong but it seems like this is kind of relevant to this suggestion from your overall comment: 

> Instead of the block diagram PDFer generating the file directly, it should just return the data. Or it takes in a FileOutputStream - so the IO exceptions become out of scope.

I will see if I can refactor the implementation of PDFGeneratorUtil to not generate the file directly.